### ISSUE TITLE
Puli filesystem repository deletes *all* symlinked files

### ISSUE THREAD

**Issue by temp** [puli/issues#178]
After a puli build I noticed that *all* of my resource files were simply deleted from my drive - luckily I had no modifications ;-)
Tried to remember what I did, and I noticed that I moved one of the resource dirs before calling puli build.
Seems like puli doesn't check the (invalid) symlinks, and instead of removing the symlink, it removes the file.

Here are the steps to reproduce:
```
mkdir test
cd test
wget https://github.com/puli/cli/releases/download/1.0.0-beta10/puli.phar
chmod +x puli.phar
mkdir res1 res2
touch res1/a res1/b res1/c res2/d res2/e res2/f
ls -al res1 res2
./puli.phar config repository.type filesystem
./puli.phar map /app1 res1
./puli.phar map /app2 res2
./puli.phar build
mv res1 resx
./puli.phar build
ls -al resx res2
```

Expected:
resx: a b c 
res2: d e f

Actual:
resx: no files
res2: no files

**Comment by tgalopin** [comment]
Are you under Windows with Cygwin or direct Linux? It could be a symlink Windows resolvment problem.

**Comment by temp** [comment]
Mac :-D
But I saw this behaviour on windows some time ago, on the windows machine of a colleague. But we weren't able to reproduce it at that time.

**Comment by webmozart** [comment]
Thanks for the reproduction code!

@tgalopin Do you have time to look at this?

**Comment by tgalopin** [comment]
@webmozart Not right now but I will :) .

**Comment by tgalopin** [comment]
```
mkdir test
cd test
wget https://github.com/puli/cli/releases/download/1.0.0-beta10/puli.phar
chmod +x puli.phar
mkdir res1 res2
touch res1/a res1/b res1/c res2/d res2/e res2/f
ls -al res1 res2
./puli.phar config repository.type filesystem
./puli.phar map /app1 res1
./puli.phar map /app2 res2
./puli.phar build
```

seems to be enough to see the problem. I'm working on it.

**Comment by tgalopin** [comment]
I found the problem, I make a PR right now.

**Comment by tgalopin** [comment]
In short: the problem comes from directory links and how `FilesystemRepository::clear` handle them.

In details:

The `build` cli command do two things: clear the repository and rebuild it. To clean the repository, the manager call the method `clear` of the configured repository (in our case FilsystemRepository).

In the FilesystemRepository, `clear` calls `removeResource` on all the resources in the repository. When using the FilesystemRepository with symlinks, it means `removeResource` is called on all the links present in `.puli/path-mappings`. However, `removeResource` is, for the moment, always recursive: all the children of the given path will be removed before trying to remove the path itself.

When trying to remove a directory link, `removeResource` list all its children to remove them: however, the children of a directory link are the real children in the filesystem. So the real files are removed.

One way to fix this would be without breaking BC would be to add an argument to `removeResource` to don't iterate recusrively on children in the context of `clear` (as we already did in the `clear` method itself). I'll create a PR to solve that.

**Comment by webmozart** [comment]
What about checking whether a file is a link before deciding to recurse into it?

**Comment by tgalopin** [comment]
I think you mean a directory? It may indeed work. I'll create a test and try to fix it using your solution in priority as it seems more logical :) .

**Comment by tgalopin** [comment]
I exposed the problem with two tests here: https://github.com/puli/repository/pull/91

**Comment by webmozart** [comment]
This is fixed in puli/repository#91.
